env-check.sh

```diff
@@ -31,19 +31,26 @@ func_check_file(){
     if [ -e "$1" ]; then
         return
     fi
-    out_str=$out_str"\nOptional: Enable dynamic debug logs on \e[31m $1 \e[0m file\n\tFor example,\n\toptions snd_sof dyndbg=+p\n\toptions snd_sof_pci dyndbg=+p"
+    out_str="$out_str""Optional: Enable dynamic debug logs in \e[31m $1 \e[0m file\n\tFor example,\n\toptions snd_sof dyndbg=+p\n\toptions snd_sof_pci dyndbg=+p\n"
     check_res=1
 }
 
 check_res=0
-echo -ne "Check for the package:\t\t"
+printf "Checking for some OS packages:\t\t"
 func_check_pkg expect
 func_check_pkg aplay
 func_check_pkg python3
 func_check_python_pkg graphviz
 func_check_file "$DYNDBG"
-[[ $check_res -eq 0 ]] && echo "pass" || \
-    echo -e "\e[31mWarning\e[0m\nSolution:"$out_str
+if [ $check_res -eq 0 ]; then
+    printf "pass\n"
+else
+    printf '\e[31mWarning\e[0m\n'
+# Need ANSI color characters to be the format string. This is not
+# unsanitized input.
+# shellcheck disable=SC2059
+    printf "$out_str"
```

Comment:
what are the advantages of using `printf` over `echo`?

Reply:
`echo` is a mess https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo/65819

Reply:
ok, good to know, thanks. Still, I think as long as we don't use sufficiently advanced "echo" features and our shell is fixed to `#!/bin/bash` in most cases `echo` should work just fine, so hopefully we aren't going to start replacing it everywhere with `printf`

Reply:
@marc-hb btw, while at it, do we really need those escape sequences for colour coding etc.? They are fun and (arguably) useful when working with a terminal, but they aren't helpful when redirecting output to a file. Ideally we'd want to check whether the output is redirected like git and others do...

Reply:
Preferably variables for the escape sequences. Yeah.

Reply:
Thanks for the reviews!

I'm not aware of any plan to mass search/replace of `echo` but I generally replace it when I'm modifying the code anyway for some other reason. It's rare to print a pure string without any variable and then it's easier to just `printf '%s' "$VAR"` rather than think about what could be in the string; and then these assumptions to become wrong in the future. While shell is admittedly safer than C, you would never `printf(some_string_variable);` in C without a format either. This would be flagged by most compilers and analyzers anyway.

I'm not a fan of the escape sequences either but I'm trying to focus on more important stuff right now :-) In this case I'm not aware of `env-check.sh` being used programmatically.

BTW you can use `less -R` to view files with escape sequences.

> Preferably variables for the escape sequences. Yeah.

I don't understand this sorry.

Reply:
I mean have variables like $RED, $RESET etc which could either contain the escape sequences or empty strings.

Reply:
There are many tasks where no other language can beat a short shell script (< few hundred lines). But for building a good, interactive user interface I would probably looks elsewhere.
tl;dr: I think the colors are overkill here, some basic "ASCII art" would be simpler and good enough.